### ISSUE TITLE
The stylesheet of a CustomizableHtmlReportImpl does not make the task outdated or cause a cache miss

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#8473]
If you change the stylesheet of an html reporting task, it is neither considered out of date, nor do you get a cache miss.
I added the SpotBugs plugin to a build and played with it.
I changed the stylesheet from `fancy-hist.xsl` to `fancy.xsl` to see the difference in the result, but the task stayed up to date.
I then deleted the report file and the task result was taken from the build cache (explicitly activated of course).
Only when I deleted the output file and run with `--no-build-cache` was the task rerun with the new stylesheet.
This does not only happen for different stylesheets, but of course also for changes in the same stylesheet.

Current workaround of course is to add the stylesheet as input manually like
```gradle
configurations { spotbugsStylesheets { transitive false } }
dependencies { spotbugsStylesheets 'com.github.spotbugs:spotbugs:3.1.11' }
tasks.withType(SpotBugsTask) {
    def spotbugsStylesheet = resources.text.fromArchiveEntry(configurations.spotbugsStylesheets, 'fancy-hist.xsl')
    inputs.file spotbugsStylesheet.asFile() withPropertyName 'spotbugsStylesheet' withPathSensitivity NONE
    reports {
        xml.enabled false
        html {
            enabled true
            stylesheet spotbugsStylesheet
        }
    }
}
```

**Comment by big-guy** [comment]
Is it possible this is a spotbugs only problem?

We have a test that appears to work that does this (switches between two stylesheets):
https://github.com/gradle/gradle/blob/1d6d36692fe79b8fe4bbf985de66e13ce80563ac/subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/findbugs/AbstractFindBugsPluginIntegrationTest.groovy#L209-L237

**Comment by Vampire** [comment]
I don't think so, I get the very same behavior when I use the FindBugs plugin.

**Comment by big-guy** [comment]
Hmm, could you reopen this ticket once you have a sample you can share? Do you see this with TextResources that come from a local file (vs `resources.text.fromArchiveEntry`)? Thanks!

**Comment by Vampire** [comment]
I'll test it, but I cannot reopen a ticket, I can only comment.
Besides that, I tried to execute the test you mentione, but wasn't able to, how do you run it?
If I run with `./gradlew codeQuality:integTest --tests 'org.gradle.api.plugins.quality.findbugs.FindBugsIntegrationTest.can generate html reports with a custom stylesheet' -i` on Java 11, the test is not run as it is marked to only run with Java 8, if I run the same with Java 8, the whole build complains, that it cannot be run with Java 8 and if I remove the Java 8 requirement the test of course fails, because the ASM version in FindBugs cannot parse the file compiled with Java 11.

**Comment by Vampire** [comment]
Ah, got it to run by removing the Java 8 requirement and adding `compileJava.options.compilerArgs << '--release' << '8'` to the testee buildfile and it worked like for you.
And you are probably right, so here you have your reproducer. Change tha code to
```groovy
def "can generate html reports with a custom stylesheet"() {
    given:
    buildFile << """
        configurations {
            findbugsStylesheets { transitive false }
        }
        dependencies {
            findbugsStylesheets 'com.google.code.findbugs:findbugs:3.0.1'
        }
        compileJava.options.compilerArgs << '--release' << '8'
        findbugsMain.reports {
            xml.enabled false
            html.enabled true
            html.stylesheet resources.text.fromArchiveEntry(configurations.findbugsStylesheets, 'fancy.xsl')
        }
    """
    goodCode()
    when:
    def result = succeeds("findbugsMain")

    then:
    result.assertTaskNotSkipped(":findbugsMain")
    file("build/reports/findbugs/main.html").exists()
    file("build/reports/findbugs/main.html").assertContents(not(containsString("History")))

    when:
    // Change to a custom stylesheet
    buildFile << """
        findbugsMain.reports {
            html.stylesheet resources.text.fromArchiveEntry(configurations.findbugsStylesheets, 'fancy-hist.xsl')
        }
    """
    result = succeeds "findbugsMain"

    then:
    result.assertTaskNotSkipped(":findbugsMain")
    file("build/reports/findbugs/main.html").exists()
    file("build/reports/findbugs/main.html").assertContents(containsString("History"))
}
```
et voilá, it fails.
On the second run the `findbugsMain` task is up-to-date.
I also added `assertTaskNotSkipped` assertions, the second one already fails before the content check, but the content check would have also failed of course.

Btw. with `resources.text.fromString` it also works fine, I just tested with the same testcase.

**Comment by big-guy** [comment]
Thanks @Vampire

**Comment by wolfs** [comment]
So it looks like the name of the entry in the archive file is not an input, the whole archive is considered an input. And since the change is using a different file from the same archive, Gradle does not detect the change.

So either we add the path of the file in the archive as an input or we make sure we snapshot the selected file in a `zipTree` instead of the backing file.

**Comment by Vampire** [comment]
Ah, I see, that makes sense somehow, but is suboptimal of course. :-D
Also, if anything else in the archive changes, doesn't mean the task needs to be out of date or a cache miss.
So yes, I guess it would be best to use the text resource contents as input, maybe even with path sensitivity none?
Because for a stylesheet it should be insignificant where the content comes from.
